forest/cli/src/cli/snapshot_cmd.rs

```diff
@@ -323,6 +349,133 @@ fn clean(config: &Config, snapshot_dir: &Option<PathBuf>, force: bool) -> anyhow
     Ok(())
 }
 
+async fn validate(
+    config: &Config,
+    validate_height: &i64,
+    snapshot: &PathBuf,
+) -> anyhow::Result<()> {
+    let confirm = Confirm::with_theme(&ColorfulTheme::default())
+        .with_prompt(format!(
+            "This will result in using approximately {} MB of data. Proceed?",
+            std::fs::metadata(snapshot)?.len() / (1024 * 1024)
+        ))
+        .default(false)
+        .interact()
+        .unwrap_or_default();
+
+    if confirm {
+        let tmp_db_path = TempDir::new()?;
+
+        let db_path = tmp_db_path.path().join(&config.chain.name);
+
+        let db = forest_db::rocks::RocksDb::open(db_path, &config.rocks_db)
+            .expect("Opening RocksDB must succeed");
+
+        let chain_store = Arc::new(ChainStore::new(db).await);
+
+        let genesis = read_genesis_header(
+            config.client.genesis_file.as_ref(),
+            config.chain.genesis_bytes(),
+            &chain_store,
+        )
+        .await
+        .unwrap();
+        let genesis_header = &genesis.blocks()[0];
+        chain_store.set_genesis(genesis_header).unwrap();
+
+        let cids = {
+            let file = tokio::fs::File::open(&snapshot).await?;
+            let reader = FetchProgress::fetch_from_file(file).await?;
+            load_car(&chain_store.db, reader.compat()).await?
+        };
+
+        let ts = chain_store.tipset_from_keys(&TipsetKeys::new(cids)).await?;
+
+        validate_links_and_genesis_traversal(
+            &chain_store,
+            ts,
+            &chain_store.db,
+            *validate_height,
+            &genesis,
+        )
+        .await?;
+    }
+
+    Ok(())
+}
+
+async fn validate_links_and_genesis_traversal<DB>(
+    chain_store: &ChainStore<DB>,
+    ts: Arc<Tipset>,
+    db: &DB,
+    upto: ChainEpoch,
+    genesis_tipset: &Tipset,
+) -> anyhow::Result<()>
+where
+    DB: fvm_ipld_blockstore::Blockstore + Store + Send + Sync + Clone,
+{
+    let mut seen = std::collections::HashSet::<Cid>::new();
+    let upto = ts.epoch() - upto;
+
+    println!(
+        "validating cids from snapshot in db: {} <- {} ...",
+        upto,
+        ts.epoch()
+    );
+
+    let mut tsk = ts.parents().clone();
+
+    // Security: Recursive snapshots are difficult to create but not impossible. This
+    // limits the amount of recursion we do.
+    let mut iteration_guard = ts.epoch() + 1;
+    let genesis_reachable = loop {
+        if iteration_guard <= 0 {
+            break false;
+        }
+
+        let tipset = chain_store.tipset_from_keys(&tsk).await?;
+        if iteration_guard - tipset.epoch() < 1 {
+            bail!(
+                "tipset iteration didn't make progress backwards, possible recursion in snapshot."
+            );
+        }
```

Comment:
I don't understand this logic. Could you explain it to me or replace it with something so simple it requires no explanation?

Reply:
We start with this invariant: `let mut iteration_guard = ts.epoch() + 1;` outside of loop; that `ts.epoch()` is always smaller than the guard by atleast 1. 

and if it violates (which can only happen when tipset.epoch()'s value didn't decrease by parent() calls), we bail.

Reply:
How about keeping track of the previous epoch and verifying the new epoch is smaller?